File: packages/ui/src/i18n.ts
/**
 * N-WP13: the whole i18n runtime — a catalogue lookup, `{placeholder}`
 * substitution, a language guess and a plural rule.
 *
 * The shape is nazar-tray's (`ui/src/i18n.ts` in that repo), adapted rather
 * than copied. What is the same, because it was right there: flat key/value
 * JSON files with `{name}` placeholders, English behind every miss, a key
 * printed as itself when nobody wrote it, and a hand-written plural rule
 * instead of `Intl.PluralRules`. What is different is below.
 *
 * **The chosen language is module state, and that is a decision.** The tray
 * threads a `Translate` through its render functions; Nazar's canvas builds
 * text in about a dozen modules and four of them are pure helpers with their
 * own unit tests (`format.ts`, `quota-strip.ts`, `hidden.ts`, `colours.ts`).
 * Threading a translator through those would have put an argument on forty
 * signatures to express one fact — *this page is in one language at a time* —
 * that is genuinely global. So the language lives here, `t()` reads it, and
 * changing it is one call followed by a redraw.
 *
 * **The catalogues are installed rather than imported.** This module is
 * compiled for the package as well as bundled for the browser, and
 * `packages/ui/tsconfig.json` roots its build at `src/`, so it cannot reach
 * `../locales`. The browser half (`web/locales.ts`) imports the six JSON files
 * and installs them before the first frame; the tests read the same files off
 * disk and install them too. With nothing installed `t()` returns the key,
 * which is loud rather than blank — and is what the parity tests assert on.
 *
 * **Flat strings only.** Every value in a locale file is a string. Not a
 * stylistic rule: `apps/desktop/src/i18n.rs` parses the same files as
 * `BTreeMap<String, String>` for the tray menu and the window title, and one
 * nested object anywhere makes the whole file fail to parse — silently, because
 * a damaged translation must not stop the shell from starting. So that language
 * would quietly fall back to English. The translation status therefore lives in
 * `locales/README.md` and not in a `_meta` key inside the files.
 *
 * **Writing direction.** All six languages v1 ships are left to right, so the
 * page sets `<html lang>` and never `dir`. Adding Arabic, Hebrew or Persian
 * means a direction table here *and* an audit of `web/styles.css`, which still
 * uses physical `margin-left` and `text-align: right` in a dozen places;
 * `i18n.test.ts` fails if an RTL tag is added to {@link LOCALES} before that
 * work is done.
 */

/** Every language v1 ships, in the order the picker lists them. */
export const LOCALES = ['en', 'tr', 'zh', 'ko', 'ru', 'es'] as const;

/** One of {@link LOCALES}. */
export type Locale = (typeof LOCALES)[number];

/** The language every missing string falls back to. */
export const FALLBACK_LOCALE: Locale = 'en';

/** Where the chosen language is kept. Absent until somebody picks one. */
export const LOCALE_KEY = 'nazar.locale.v1';

/** A flat map of message key to message template. */
export type Catalog = Readonly<Record<string, string>>;

/** Every catalogue, keyed by language. */
export type Catalogs = Readonly<Record<Locale, Catalog>>;

/** Values substituted into `{placeholder}` slots. */
export type Params = Readonly<Record<string, string | number>>;

/** Narrow an arbitrary string to a language this build can paint itself in. */
export function isLocale(value: string): value is Locale {
  return (LOCALES as readonly string[]).includes(value);
}

/** The primary subtag of a language tag: `tr-TR` and `TR_tr` both become `tr`. */
export function primarySubtag(tag: string): string {
  return (tag.split(/[-_]/)[0] ?? '').toLowerCase();
}

/**
 * Replace `{name}` with `params.name`.
 *
 * A placeholder with no matching parameter is left as written rather than
 * blanked, so a missing value shows up as `{count}` on screen instead of
 * disappearing quietly.
 */
export function interpolate(template: string, params?: Params): string {
  if (params === undefined) return template;
  return template.replace(/\{(\w+)\}/g, (whole, name: string) => {
    const value = params[name];
    return value === undefined ? whole : String(value);
  });
}

/* ------------------------------------------------------------------ *
 * Plurals
 * ------------------------------------------------------------------ */

/** The plural category a count falls into, for the languages this product ships. */
export type PluralCategory = 'one' | 'few' | 'other';

/**
 * The plural category of `count` in `locale`.
 *
 * Three of the six need no rule at all — Chinese, Korean and Turkish leave a
 * noun alone after a numeral — English and Spanish need two forms, and Russian
 * needs the three the grammar books give: **1** (`one`), **2–4** (`few`), **5
 * and up** (`other`), with the teens 11–14 taken out of both, so 21 is `one`
 * and 111 is `other`.
 *
 * `Intl.PluralRules` would answer the same question and is deliberately not
 * used: it would put a shipped string at the mercy of the WebView2 version on
 * the user's machine, and `apps/desktop/src/i18n.rs` has no `Intl` to agree
 * with. Fifteen lines both halves can read is the cheaper contract.
 */
export function pluralCategory(count: number, locale: Locale): PluralCategory {
  const n = Math.abs(Math.trunc(count));
  if (locale === 'ru') {
    const tens = n % 100;
    const units = n % 10;
    if (tens >= 11 && tens <= 14) return 'other';
    if (units === 1) return 'one';
    if (units >= 2 && units <= 4) return 'few';
    return 'other';
  }
  if (locale === 'en' || locale === 'es') return n === 1 ? 'one' : 'other';
  return 'other';
}

/**
 * The message key for a counted word, chosen by the count.
 *
 * Nazar has real counted **words** where the tray had only unit abbreviations —
 * *3 sessions*, *1 subagent*, *2 finished subagents hidden* — so unlike the
 * tray this helper is used, and used from the first day. A key `x.sessions`
 * becomes `x.sessions.one`, `x.sessions.few` or `x.sessions.other`, and every
 * catalogue carries all three for every counted key: `i18n.test.ts` fails on a
 * key whose forms are not complete, in any language, including the ones whose
 * three forms are the same word.
 */
export function pluralKey(key: string, count: number, locale: Locale): string {
  return `${key}.${pluralCategory(count, locale)}`;
}

/* ------------------------------------------------------------------ *
 * The chosen language
 * ------------------------------------------------------------------ */

const EMPTY: Catalog = {};

let catalogs: Catalogs | undefined;
let locale: Locale = FALLBACK_LOCALE;

/** Bind the six catalogues. Called once, before the first frame. */
export function installCatalogs(next: Catalogs): void {
  catalogs = next;
}

/** The language in force. */
export function currentLocale(): Locale {
  return locale;
}

/** Switch language. `true` when it actually moved, which is what asks for a redraw. */
export function setLocale(next: Locale): boolean {
  if (next === locale) return false;
  locale = next;
  return true;
}

/**
 * The language the page is in: the stored choice, then the browser's list, then English.
 *
 * A tag no catalogue answers for falls through rather than being selected, so a
 * machine set to German gets an English page instead of a page full of message
 * keys. Only the primary subtag is compared: there is one Spanish here, not
 * `es-MX` and `es-ES`.
 */
export function resolveLocale(
  chosen: string | undefined,
  preferred: readonly string[] = [],
): Locale {
  for (const tag of chosen === undefined ? preferred : [chosen, ...preferred]) {
    const primary = primarySubtag(tag);
    if (isLocale(primary)) return primary;
  }
  return FALLBACK_LOCALE;
}

/**
 * The message for a key, with its placeholders filled in.
 *
 * Lookup order is the chosen language, then English, then the key itself.
 * Returning the key makes an untranslated string obvious on screen — a menu
 * entry reading `menu.jump` — instead of rendering as a gap nobody notices.
 */
export function t(key: string, params?: Params): string {
  const chosen = catalogs?.[locale] ?? EMPTY;
  const fallback = catalogs?.[FALLBACK_LOCALE] ?? EMPTY;
  return interpolate(chosen[key] ?? fallback[key] ?? key, params);
}

/** {@link t}, for a counted word: picks the form and fills `{count}` in. */
export function tCount(key: string, count: number, params?: Params): string {
  return t(pluralKey(key, count, locale), { count, ...params });
}

/* ------------------------------------------------------------------ *
 * Storage
 * ------------------------------------------------------------------ */

/** The sliver of `localStorage` this module uses. Mirrors `workspace.ts`. */
export interface StorageLike {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
  removeItem(key: string): void;
}

/**
 * The stored choice, or `undefined` when nobody has picked a language yet.
 *
 * Only the primary subtag is ever stored, and **anything this build cannot
 * paint itself in is read as no choice at all** rather than kept. That covers
 * two cases with one rule: a language dropped from a later version must not
 * leave somebody stuck on message keys, and a build before N-WP13b could write
 * the word `system` here — the picker offered *follow the machine* as if it
 * were a seventh language. It no longer does, and the value it left behind
 * means what it always meant: resolve the language from the machine.
 */
export function readLocaleChoice(storage: StorageLike | undefined): Locale | undefined {
  if (storage === undefined) return undefined;
  const raw = storage.getItem(LOCALE_KEY);
  if (raw === null) return undefined;
  const primary = primarySubtag(raw);
  return isLocale(primary) ? primary : undefined;
}

/**
 * Store the picked language.
 *
 * There is no way back to *follow the machine* because there is nothing to go
 * back to: following the machine is what an empty key already means, and the
 * picker has no entry that would clear it.
 */
export function writeLocaleChoice(storage: StorageLike | undefined, choice: Locale): void {
  if (storage === undefined) return;
  try {
    storage.setItem(LOCALE_KEY, choice);
  } catch {
    // Storage can be blocked. The choice simply does not survive a reload.
  }
}

File: apps/desktop/src/i18n.rs
//! N-WP13: the few strings the shell draws, out of the same files the canvas reads.
//!
//! The tray menu and its tooltip are the only text this application puts on screen
//! outside the webview, and the rule the canvas keeps — *every visible string comes from
//! a locale file* — has no exception for them. So `packages/ui/locales/<lang>.json` is
//! compiled in with `include_str!` and looked up with the same `{placeholder}` rule
//! `packages/ui/src/i18n.ts` uses, in about sixty lines. The shape is nazar-tray's
//! `crates/nazar-tray/src/i18n.rs`, adapted.
//!
//! **The shell does not know what the canvas chose, and does not ask.** The canvas keeps
//! its language in `localStorage`, which belongs to the page; the tray menu is drawn
//! before any page exists and by a process that has no access to it. So the shell
//! resolves the *system* UI language and uses that ([`resolve`]). The two can therefore
//! disagree — a Turkish machine whose canvas has been switched to English gets an English
//! canvas under a Turkish tray — and that is the honest trade for not inventing an IPC
//! call and a settings file to carry one enum. `docs/PROJECT.md` §7 records it.
//!
//! **Every value in a locale file must be a string, and the price of forgetting is
//! silence.** [`parse`] asks serde for a `BTreeMap<String, String>`; a file with one
//! nested object in it does not parse, becomes an empty catalogue, and that language
//! quietly falls back to English. It is the right failure — a damaged translation must
//! not stop the shell from starting — but it is an invisible one, which is why
//! `packages/ui/test/i18n.test.ts` fails on a non-string value.

use std::collections::BTreeMap;

/// English, the fallback for every missing string.
const EN: &str = include_str!("../../../packages/ui/locales/en.json");
/// Turkish, written by hand alongside English.
const TR: &str = include_str!("../../../packages/ui/locales/tr.json");
/// The four that were machine-translated first; corrections arrive by pull request.
const ZH: &str = include_str!("../../../packages/ui/locales/zh.json");
const KO: &str = include_str!("../../../packages/ui/locales/ko.json");
const RU: &str = include_str!("../../../packages/ui/locales/ru.json");
const ES: &str = include_str!("../../../packages/ui/locales/es.json");

/// Every language this build can paint itself in, in the order the canvas lists them.
pub const LOCALES: [&str; 6] = ["en", "tr", "zh", "ko", "ru", "es"];

/// A catalogue bound to one language, with English behind it.
#[derive(Debug)]
pub struct Catalog {
    messages: BTreeMap<String, String>,
    fallback: BTreeMap<String, String>,
}

/// Parse one catalogue. A file that does not parse is an empty catalogue, never a panic.
fn parse(text: &str) -> BTreeMap<String, String> {
    serde_json::from_str(text).unwrap_or_default()
}

/// The primary subtag of a language tag: `tr-TR` and `TR_tr` both become `tr`.
#[must_use]
pub fn primary_subtag(locale: &str) -> String {
    locale
        .split(['-', '_'])
        .next()
        .unwrap_or_default()
        .to_ascii_lowercase()
}

/// The source text of a catalogue this build carries, or `None`.
fn source(primary: &str) -> Option<&'static str> {
    match primary {
        "en" => Some(EN),
        "tr" => Some(TR),
        "zh" => Some(ZH),
        "ko" => Some(KO),
        "ru" => Some(RU),
        "es" => Some(ES),
        _ => None,
    }
}

/// The catalogue for a language tag such as `tr`, `tr-TR` or `en-GB`.
#[must_use]
pub fn catalog(locale: &str) -> Catalog {
    let primary = primary_subtag(locale);
    let messages = match primary.as_str() {
        // English is the fallback; loading it twice would only double the memory.
        "en" => BTreeMap::new(),
        other => source(other).map(parse).unwrap_or_default(),
    };
    Catalog {
        messages,
        fallback: parse(EN),
    }
}

/// The language the shell paints itself in: the machine's, then English.
///
/// A tag no catalogue answers for falls through rather than being selected, so a machine
/// set to German gets an English tray instead of one full of message keys.
#[must_use]
pub fn resolve(system: Option<&str>) -> String {
    if let Some(tag) = system {
        let primary = primary_subtag(tag);
        if LOCALES.contains(&primary.as_str()) {
            return primary;
        }
    }
    "en".to_owned()
}

impl Catalog {
    /// The message for a key, with the placeholders filled in.
    ///
    /// Lookup order is the chosen language, then English, then the key itself — an
    /// untranslated string shows up as `tray.quit` in the menu instead of as a gap.
    #[must_use]
    pub fn format(&self, key: &str, params: &[(&str, &str)]) -> String {
        let template = self
            .messages
            .get(key)
            .or_else(|| self.fallback.get(key))
            .map_or(key, String::as_str);
        interpolate(template, params)
    }

    /// The message for a key that has no placeholders.
    #[must_use]
    pub fn text(&self, key: &str) -> String {
        self.format(key, &[])
    }

    /// Every key English defines. Used by the parity tests.
    #[cfg(test)]
    #[must_use]
    pub fn keys(&self) -> Vec<&str> {
        self.fallback.keys().map(String::as_str).collect()
    }

    /// Whether the chosen language defines a key of its own.
    #[cfg(test)]
    #[must_use]
    pub fn translates(&self, key: &str) -> bool {
        self.messages.contains_key(key)
    }
}

/// Replace `{name}` with the value given for `name`.
///
/// A placeholder with no value is left as written rather than blanked, so a missing value
/// reads as `{percent}` instead of disappearing — the same rule as `src/i18n.ts`.
#[must_use]
fn interpolate(template: &str, params: &[(&str, &str)]) -> String {
    let mut out = String::with_capacity(template.len());
    let mut rest = template;
    while let Some(start) = rest.find('{') {
        let Some(end) = rest[start..].find('}').map(|at| start + at) else {
            break;
        };
        let name = &rest[start + 1..end];
        out.push_str(&rest[..start]);
        match params.iter().find(|(key, _)| *key == name) {
            Some((_, value)) => out.push_str(value),
            None => out.push_str(&rest[start..=end]),
        }
        rest = &rest[end + 1..];
    }
    out.push_str(rest);
    out
}

/// The operating system's UI language, as a tag, or `None`.
///
/// Windows answers through `GetUserDefaultLocaleName`; everywhere else the POSIX
/// environment is the convention and is what every other tool on the machine reads.
/// `C` and `POSIX` mean *no preference* rather than a language, so they fall through.
#[must_use]
pub fn system_language() -> Option<String> {
    #[cfg(target_os = "windows")]
    {
        if let Some(tag) = windows_ui_language() {
            return Some(tag);
        }
    }
    for name in ["LC_ALL", "LC_MESSAGES", "LANG"] {
        let Ok(value) = std::env::var(name) else {
            continue;
        };
        let tag = value.split('.').next().unwrap_or_default();
        if tag.is_empty() || tag == "C" || tag == "POSIX" {
            continue;
        }
        return Some(tag.to_owned());
    }
    None
}

#[cfg(target_os = "windows")]
fn windows_ui_language() -> Option<String> {
    use windows::Win32::Globalization::GetUserDefaultLocaleName;
    // The documented maximum is 85 UTF-16 code units including the terminator.
    let mut buffer = [0u16; 85];
    // SAFETY: the wrapper hands the slice and its length over together, and the call
    // writes at most that many units into it.
    let written = unsafe { GetUserDefaultLocaleName(&mut buffer) };
    if written <= 1 {
        return None;
    }
    // The count includes the terminating null.
    String::from_utf16(&buffer[..(written as usize - 1)]).ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_placeholder_is_filled_and_an_unknown_one_is_left_alone() {
        assert_eq!(
            interpolate("{a} and {b}", &[("a", "1"), ("b", "2")]),
            "1 and 2"
        );
        assert_eq!(
            interpolate("{a} and {b}", &[("a", "1")]),
            "1 and {b}",
            "a missing value must be visible, not silently blank"
        );
        assert_eq!(interpolate("nothing to fill", &[]), "nothing to fill");
        assert_eq!(interpolate("{unclosed", &[]), "{unclosed");
    }

    #[test]
    fn each_language_answers_for_itself_and_falls_back_to_english() {
        assert_eq!(catalog("en").text("tray.quit"), "Quit");
        assert_eq!(catalog("tr-TR").text("tray.quit"), "Çık");
        assert_eq!(catalog("zh-CN").text("tray.quit"), "退出");
        assert_eq!(catalog("ko").text("tray.quit"), "종료");
        assert_eq!(catalog("ru").text("tray.quit"), "Выйти");
        assert_eq!(catalog("es-MX").text("tray.quit"), "Salir");
        assert_eq!(
            catalog("de-DE").text("tray.quit"),
            "Quit",
            "a language nobody has translated gets English, not message keys"
        );
    }

    #[test]
    fn a_key_nobody_wrote_shows_as_the_key() {
        assert_eq!(catalog("en").text("nothing.like.this"), "nothing.like.this");
    }

    #[test]
    fn every_english_key_is_translated_into_every_other_language() {
        for locale in LOCALES {
            if locale == "en" {
                continue;
            }
            let catalog = catalog(locale);
            let missing: Vec<&str> = catalog
                .keys()
                .into_iter()
                .filter(|key| !catalog.translates(key))
                .collect();
            assert!(
                missing.is_empty(),
                "{locale} is behind English on {} keys: {missing:?}",
                missing.len()
            );
        }
    }

    #[test]
    fn every_key_the_shell_draws_exists_in_all_six() {
        // Read out of this crate's own sources, so a new `catalog.text(…)` is covered
        // the moment it is written rather than the next time somebody remembers.
        let source_dir = std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("src");
        let mut keys: Vec<String> = Vec::new();
        for entry in std::fs::read_dir(&source_dir).expect("the crate has a src directory") {
            let path = entry.expect("a readable entry").path();
            if path.extension().and_then(|e| e.to_str()) != Some("rs") {
                continue;
            }
            let Ok(text) = std::fs::read_to_string(&path) else {
                continue;
            };
            let mut rest = text.as_str();
            while let Some(at) = rest.find(".text(\"") {
                rest = &rest[at + 7..];
                if let Some(end) = rest.find('"') {
                    keys.push(rest[..end].to_owned());
                }
            }
        }
        keys.retain(|key| key != "nothing.like.this");
        keys.sort();
        keys.dedup();
        assert!(
            keys.len() >= 4,
            "the scan found only {} keys, so it has stopped working: {keys:?}",
            keys.len()
        );
        for locale in LOCALES {
            let catalog = catalog(locale);
            for key in &keys {
                assert!(
                    locale == "en" || catalog.translates(key),
                    "{locale} does not translate {key}, which the shell draws"
                );
            }
        }
    }

    #[test]
    fn the_language_is_the_machines_and_then_english() {
        assert_eq!(resolve(Some("tr-TR")), "tr");
        assert_eq!(resolve(Some("TR_tr")), "tr", "tags are normalised");
        assert_eq!(resolve(None), "en", "English is the floor");
        assert_eq!(
            resolve(Some("de-DE")),
            "en",
            "an English tray beats a tray full of message keys"
        );
        for locale in LOCALES {
            assert_eq!(resolve(Some(locale)), locale);
        }
    }
}
